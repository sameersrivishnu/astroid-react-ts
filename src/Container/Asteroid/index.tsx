import React, { useState, useCallback } from "react";
import {
  Box,
  TextField,
  Button,
  Typography,
  CircularProgress
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import API from "./../../API/API";
import { API_KEY } from "./../../Helpers/Constants";
// API
const api = new API();

const useStyles = makeStyles(theme => ({
  searchContainer: {
    border: "1px solid #e2d2d2",
    padding: theme.spacing(3, 2)
  },
  buttonContainer: {
    margin: theme.spacing(1, 0),
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between"
  },
  infoContainer: {
    color: theme.palette.success.main
  },
  detailsWrapper: {
    marginTop: theme.spacing(1),
    border: "1px solid #e2d2d2",
    padding: theme.spacing(3, 2)
  },
  itemContainer: {
    marginBottom: theme.spacing(1),
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-start",
    paddingBottom: theme.spacing(1),
    borderBottom: "1px solid #e2e2e2"
  },
  titleContainer: {
    width: "25%"
  },
  loaderContainer: {
    textAlign: "center"
  }
}));

/**
 * Get Asteroid base
 */
export default function TopBar() {
  const classes = useStyles();
  const [asteroidId, setAsteroidId] = useState("");
  const [asteroidData, setAsteroidData] = useState(null as any);
  const [isBusy, setIsBusy] = useState(false);

  const handleChange = useCallback(e => {
    setAsteroidId(e.target.value);
  }, []);

  const searchAsteroid = useCallback(() => {
    if (asteroidId) {
      console.log("searchAsteroid -> asteroidId", asteroidId);
      setIsBusy(true);
      api.get(`neo/${asteroidId}?api_key=${API_KEY}`).then(response => {
        console.log("searchAsteroid -> response", response);
        setIsBusy(false);
        if (response) {
          setAsteroidData(response);
        }
      });
    }
  }, [asteroidId]);

  const getRandomAsteroidData = useCallback(() => {
    setIsBusy(true);
    api.get(`neo/browse?api_key=${API_KEY}`).then(response => {
      setIsBusy(false);
      if (
        response &&
        response.near_earth_objects &&
        response.near_earth_objects.length
      ) {
        const asteroidId = Math.floor(Math.random() * 20);
        setAsteroidData(response.near_earth_objects[asteroidId]);
        setAsteroidId(response.near_earth_objects[asteroidId].id);
      }
    });
  }, []);

  return (
    <Box style={{ width: '99%', marginLeft: 7 }}>
      <Box className={classes.searchContainer}>
        <TextField
          id="outlined-basic"
          label="Enter Asteroid Id"
          variant="outlined"
          onChange={handleChange}
          fullWidth
          value={asteroidId}
        />
        <Box className={classes.buttonContainer}>
          <Button
            variant="contained"
            color="secondary"
            onClick={searchAsteroid}
          >
            Search
          </Button>
          <Button
            variant="contained"
            color="primary"
            onClick={getRandomAsteroidData}
          >
            Random
          </Button>
        </Box>
      </Box>
      <Box className={classes.detailsWrapper}>
        {isBusy ? (
          <Box className={classes.loaderContainer}>
            <CircularProgress color="secondary" />
          </Box>
        ) : (
          <React.Fragment>
            {asteroidData ? (
              <Box>
                <Box className={classes.infoContainer}>
                  <Typography variant="h6" gutterBottom>
                    Successfully got the asteroid data.
                  </Typography>
                </Box>
                <Box className={classes.itemContainer}>
                  <Box className={classes.titleContainer}>ID:....</Box>
                  <Box>{asteroidData.id}</Box>
                </Box>
                <Box className={classes.itemContainer}>
                  <Box className={classes.titleContainer}>Name:</Box>
                  <Box>{asteroidData.name}</Box>
                </Box>
                <Box className={classes.itemContainer}>
                  <Box className={classes.titleContainer}>NASA JPL URL:</Box>
                  <Box>{asteroidData.nasa_jpl_url}</Box>
                </Box>
                <Box className={classes.itemContainer}>
                  <Box className={classes.titleContainer}>
                    Potentially Hazardous Asteroid:
                  </Box>
                  <Box>
                    {asteroidData.is_potentially_hazardous_asteroid
                      ? "Yes"
                      : "No"}
                  </Box>
                </Box>
              </Box>
            ) : (
              <Box>
                Please search with Asteroid Id or Click on Random.
              </Box>
            )}
          </React.Fragment>
        )}
      </Box>
    </Box>
  );
}
