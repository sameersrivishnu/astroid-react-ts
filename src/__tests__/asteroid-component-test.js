import React from "react";
import Asteroid from "../Container/Asteroid";
import renderer from "react-test-renderer";

describe("Asteroid Component", () => {
  it("should render without throwing an error", async () => {
    const rendered = renderer.create(<Asteroid />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
