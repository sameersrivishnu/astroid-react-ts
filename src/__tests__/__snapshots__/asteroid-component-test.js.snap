// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asteroid Component should render without throwing an error 1`] = `
<div
  className="MuiBox-root MuiBox-root-8"
>
  <div
    className="MuiBox-root MuiBox-root-9 makeStyles-searchContainer-1"
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="outlined-basic"
        id="outlined-basic-label"
      >
        Enter Asteroid Id
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="outlined-basic"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="3542519"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-68 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-70 PrivateNotchedOutline-legendNotched-71"
          >
            <span>
              Enter Asteroid Id
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiBox-root MuiBox-root-72 makeStyles-buttonContainer-2"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Search
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Random
        </span>
      </button>
    </div>
  </div>
  <div
    className="MuiBox-root MuiBox-root-105 makeStyles-detailsWrapper-4"
  >
    <div
      className="MuiBox-root MuiBox-root-106"
    >
      No Data Found. Please search again
    </div>
  </div>
</div>
`;
